dropshot/src/handler.rs

```diff
@@ -1436,6 +1436,172 @@ impl From<HttpResponseUpdatedNoContent> for HttpHandlerResult {
+     *
+     * We just use `String`.  This might contain values that aren't valid in
+     * HTTP response headers.  But we can at least validate that at runtime, and
+     * it sure is easier to implement!
+     */
+    location: String,
+}
+
+/** See `http_response_found()` */
+pub type HttpResponseFound =
+    HttpResponseHeaders<HttpResponseFoundStatus, RedirectHeaders>;
+
+/**
+ * `http_response_found` returns an HTTP 302 "Found" response with no response
+ * body.
+ *
+ * The sole argument will become the value of the `Location` header.  This is
+ * where you want to redirect the client to.
+ *
+ * Per MDN and RFC 9110 S15.4.3, you might want to use 307 ("Temporary
+ * Redirect") or 303 ("See Other") instead.
+ */
+pub fn http_response_found(
+    location: String,
+) -> Result<HttpResponseFound, HttpError> {
+    let _ = http::HeaderValue::from_str(&location)
+        .map_err(|e| http_redirect_error(e, &location))?;
+    Ok(HttpResponseHeaders::new(
+        HttpResponseFoundStatus,
+        RedirectHeaders { location },
+    ))
+}
+
+fn http_redirect_error(
+    error: http::header::InvalidHeaderValue,
+    location: &str,
+) -> HttpError {
+    HttpError::for_internal_error(format!(
+        "error encoding redirect URL {:?}: {:#}",
+        location, error
+    ))
+}
+
+/**
+ * This internal type impls HttpCodedResponse.  External clients should use
+ * `HttpResponseFound` instead, which includes metadata about the `Location`
+ * header.
+ */
+#[doc(hidden)]
+pub struct HttpResponseFoundStatus;
+impl HttpCodedResponse for HttpResponseFoundStatus {
+    type Body = Empty;
+    const STATUS_CODE: StatusCode = StatusCode::FOUND;
+    const DESCRIPTION: &'static str = "redirect (302)";
+}
+impl From<HttpResponseFoundStatus> for HttpHandlerResult {
+    fn from(_: HttpResponseFoundStatus) -> HttpHandlerResult {
+        HttpResponseFoundStatus::for_object(Empty)
+    }
+}
+
+/** See `http_response_see_other()` */
+pub type HttpResponseSeeOther =
+    HttpResponseHeaders<HttpResponseSeeOtherStatus, RedirectHeaders>;
+
+/**
+ * `http_response_see_other` returns an HTTP 303 "See Other" response with no
+ * response body.
+ *
+ * The sole argument will become the value of the `Location` header.  This is
+ * where you want to redirect the client to.
+ *
+ * Use this (as opposed to 307 "Temporary Redirect") when you want the client to
+ * follow up with a GET, rather than whatever method they used to make the
+ * current request.  This is intended to be used after a PUT or POST to show a
+ * confirmation page or the like.
+ */
+pub fn http_response_see_other(
+    location: String,
+) -> Result<HttpResponseSeeOther, HttpError> {
+    let _ = http::HeaderValue::from_str(&location)
+        .map_err(|e| http_redirect_error(e, &location))?;
+    Ok(HttpResponseHeaders::new(
+        HttpResponseSeeOtherStatus,
+        RedirectHeaders { location },
+    ))
+}
+
+/**
+ * This internal type impls HttpCodedResponse.  External clients should use
```

Comment:
maybe as above re: "External clients"?

Reply:
Done in 74d5e18 (and updated the others too).